test-utils/transaction-factory/src/lib.rs

```diff
@@ -0,0 +1,204 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Simple transaction factory which distributes tokens from a master
+//! account to a specified number of newly created accounts.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+#![cfg_attr(not(feature = "std"), feature(alloc))]
+
+use std::collections::HashMap;
+use std::sync::Arc;
+use std::ops::Mul;
+use std::fmt::Display;
+
+use log::info;
+
+use client::block_builder::api::BlockBuilder;
+use client::runtime_api::ConstructRuntimeApi;
+use consensus_common::{BlockOrigin, ImportBlock, ForkChoiceStrategy, SelectChain};
+use consensus_common::block_import::BlockImport;
+use parity_codec::{Decode, Encode};
+use serde::Serialize;
+use sr_primitives::traits::{As, Block, Header, ProvideRuntimeApi, SimpleArithmetic};
+use substrate_service::{FactoryBlock, FactoryFullConfiguration, FullClient, new_client, ServiceFactory, ComponentClient, FullComponents};
+pub use crate::modes::Mode;
+
+mod modes;
+mod complex_mode;
+mod simple_modes;
+
+pub trait RuntimeAdapter {
+	type AccountId;
+	type Extrinsic;
+	type Balance;
+	type Moment;
+	type BlockNumber;
+	type BlockId;
+	type Hash;
+	type Secret;
+	type Index;
+	type Phase;
+	type Header;
+
+	fn transfer_extrinsic(
+		sender: &Self::AccountId,
+		key: &Self::Secret,
+		destination: &Self::AccountId,
+		amount: Self::Balance,
+		index: Self::Index,
+		phase: Self::Phase,
+		prior_block_hash: &Self::Hash,
+	) -> Self::Extrinsic;
+
+	fn timestamp_inherent(
+		ts: Self::Moment,
+		key: Self::Secret,
+		phase: Self::Phase,
+		prior_block_hash: &Self::Hash
+	) -> Self::Extrinsic;
+
+	fn minimum_balance() -> Self::Balance;
+	fn minimum_period() -> Self::Moment;
+	fn master_account_id() -> Self::AccountId;
+	fn master_account_secret() -> Self::Secret;
+	fn extract_timestamp(block_hash: Self::Hash) -> Self::Moment;
+	fn extract_index(account_id: Self::AccountId, block_hash: Self::Hash) -> Self::Index;
+	fn extract_phase(block_hash: Self::Hash) -> Self::Phase;
+	fn gen_random_account_id(seed: u64) -> Self::AccountId;
+	fn gen_random_account_secret(seed: u64) -> Self::Secret;
+}
+
+pub struct FactoryState {
+	pub block_no: u64,
+	block_in_round: u64,
+	mode: Mode,
+	rounds: u64,
+	num: u64,
+	index: u64,
+	phase: u64,
+	round: u64,
+	start_number: u64,
+}
+
+/// Manufactures transactions. The exact amount depends on
+/// `mode`, `num` and `rounds`.
+pub fn factory<F, RA>(
+	mut config: FactoryFullConfiguration<F>,
+	mode: Mode,
+	num: u64,
+	rounds: u64,
+) -> cli::error::Result<()>
+where
+	F: ServiceFactory,
+	F::RuntimeApi: ConstructRuntimeApi<FactoryBlock<F>, FullClient<F>>,
+	FullClient<F>: ProvideRuntimeApi,
+	<FullClient<F> as ProvideRuntimeApi>::Api: BlockBuilder<FactoryBlock<F>>,
+
+	RA: RuntimeAdapter,
+	<RA as RuntimeAdapter>::AccountId: Display,
```

Comment:
Why not move all these bounds into the trait?

Reply:
@bkchr I agree.  I have avoided placing bounds in traits in the past, but it has led to extremely long type signatures and confusing error messages.